web/tests/hotkey.test.js

```diff
@@ -168,19 +169,47 @@ run_test("mappings", () => {
     assert.equal(map_down(13).name, "enter");
     assert.equal(map_down(46).name, "delete");
     assert.equal(map_down(13, true).name, "enter");
-
-    assert.equal(map_press(47).name, "search"); // slash
-    assert.equal(map_press(106).name, "vim_down"); // j
-
     assert.equal(map_down(219, false, true).name, "escape"); // Ctrl + [
     assert.equal(map_down(67, false, true).name, "copy_with_c"); // Ctrl + C
     assert.equal(map_down(75, false, true).name, "search_with_k"); // Ctrl + K
     assert.equal(map_down(83, false, true).name, "star_message"); // Ctrl + S
     assert.equal(map_down(190, false, true).name, "narrow_to_compose_target"); // Ctrl + .
 
+    // Test keypress events
+    assert.equal(map_press("/", "Slash").name, "search"); // slash
+    assert.equal(map_press("j", "KeyJ").name, "vim_down"); // j
+
+    // Test ignoring shiftKey when the keyboard is probably Latin.
+    assert.equal(map_press("r", "KeyR", false).name, "reply_message"); // r
+    assert.equal(map_press("R", "KeyR", true).name, "respond_to_author"); // R
+    assert.equal(map_press("r", "KeyR", true).name, "reply_message"); // R
+    assert.equal(map_press("R", "KeyR", false).name, "respond_to_author"); // R
+
+    // Test Cyrillic:
+    assert.equal(map_press("К", "KeyR", true).name, "respond_to_author"); // R
+    assert.equal(map_press("к", "KeyR", false).name, "reply_message"); // r
+    assert.equal(map_press("/", "Slash").name, "search"); // slash
+    assert.equal(map_press("j", "KeyJ").name, "vim_down"); // j
+
+    // Test ignoring shiftKey when the keyboard is probably Latin.
+    assert.equal(map_press("r", "KeyR", false).name, "reply_message"); // r
+    assert.equal(map_press("R", "KeyR", true).name, "respond_to_author"); // R
+    assert.equal(map_press("r", "KeyR", true).name, "reply_message"); // R
+    assert.equal(map_press("R", "KeyR", false).name, "respond_to_author"); // R
+
+    // Test Cyrillic:
+    assert.equal(map_press("К", "KeyR", true).name, "respond_to_author"); // R
+    assert.equal(map_press("к", "KeyR", false).name, "reply_message"); // r
+    assert.equal(map_press("Ө", "Period", true).name, "compose_quote_reply"); // >
+    assert.equal(map_press("ө", "Period", false), undefined); // .
```

Comment:
What source did you use for the information that the key with code "Period" would on a Cyrillic layout have the key values "өӨ"?

Looking at [Wikipedia's list](https://en.wikipedia.org/wiki/JCUKEN) of Cyrillic layouts, those are not the values found on the standard Russian-language keyboard, nor on most other Cyrillic layouts including those for Ukrainian or Kazakh. The one layout I see there that has those values on that key is an [Azerbaijani](https://en.wikipedia.org/wiki/JCUKEN#Azerbaijani) layout. The Azerbaijani language is [mainly written](https://en.wikipedia.org/wiki/Azerbaijani_alphabet) in Latin script, not Cyrillic, so this is a quite uncommon layout.

The difference doesn't happen to affect the behavior of this test case. But I think it illustrates that it would be good to use a systematic, external, source for the information these tests need about various example layouts.

I took a quick look around, and I don't see a real easy way to get such data in a way you can use directly in code. So I won't insist on that.

But a bare minimum would be, for the tests that are using some kind of knowledge of how a particular layout works, to have a comment linking to the source where you got that information. This lets the reader efficiently look it up for themself.

In addition to Wikipedia, one very readable source I found from a quick google of [keyboard layout data] is http://kbdlayout.info/. Take a look at their Russian layout:
http://kbdlayout.info/kbdru
or French:
http://kbdlayout.info/kbdfrna

Reply:
I see. I am not much familiar with the culture of usage on particular keyboard layouts and was just changing to different layouts and testing. I found at least one layout that was Cyrillic (yes, it was Azerbaijani) and had this quirk so the code-comment isn't incorrect, but I don't think it matters as far as testing is concerned either way;

If even one layout has such a structure where a key commonly assigned to a symbol can be assigned to some alphabet or some other symbol, this code tests that we behave as expected for that.

![image](https://user-images.githubusercontent.com/8033238/230712856-1fb91bc2-fc3d-48d8-9e72-01d84a724cde.png)


I could change the test or elaborate on which specific keyboard I used to create that test, yes.

Reply:
> I am not much familiar with the culture of usage on particular keyboard layouts and was just changing to different layouts and testing. I found at least one layout that was Cyrillic (yes, it was Azerbaijani) and had this quirk

If you don't know the name "Cyrillic", Wikipedia has your back:
https://en.wikipedia.org/wiki/Cyrillic_script
It's the script best known for being used to write Russian.

Our tests should be based on scenarios our users are likely to see. When writing test cases for how we handle different keyboard layouts, the tests should be based on keyboard layouts that users are likely to use — not on whatever layout for a given script comes first in a list because it's for a language whose name starts with an A.

For a keyboard layout for Cyrillic, the obvious choice is a Russian layout. If you want to do something different, you can take a look at the dialog box for changing your UI language in Zulip, and look at the available languages whose names are written in Cyrillic — that gives an idea of what languages a significant number of people are actually using Zulip with.